### ISSUE TITLE
PSurfaceJOGL dispose()s PApplet twice

### ISSUE THREAD

**Issue by jdf** [processing/processing#4381]
https://github.com/processing/processing/blob/master/core/src/processing/opengl/PSurfaceJOGL.java#L872

https://github.com/processing/processing/blob/master/core/src/processing/opengl/PSurfaceJOGL.java#L877

**Comment by codeanticode** [comment]
Thanks for pointing this out. However, I have been running some tests, and seems that these lines are never reached when closing a running sketch, either by clicking the stop button in the PDE, or closing the sketch window. Are these causing you trouble in the Python mode? Maybe related to https://github.com/jdf/processing.py/issues/197#event-609204151?

**Comment by jdf** [comment]
No trouble at all, other than having to write some guard code to prevent dispose from being called twice in Python Mode.

Because the sketch running process does not exit, dispose is actually called in Python Mode. I guess Java mode quits before any dispose happens. That's actually a bug, as far far as I'm concerned, in the case where someone has written a dispose handler that cleans up temp files or whatever. It will never happen.

**Comment by leadedge** [comment]
Actually this could be related to to a problem I have been having with development of a library.

https://github.com/leadedge/SpoutProcessing

If I register a "dispose" method as indicated :

https://github.com/processing/processing/wiki/Library-Basics

it apparently never gets called. 

I have noticed this before when I used an over-ride of the "Exit" method within the sketch itself. In Processing versions before 3.0.2, the exit method was called OK if the sketch window was closed with the [X] button on the window, but not if the Stop button was used in the PDE window. 

At the time I confirmed this behaviour with a dispose method within the Java file. Now though, with the latest Processing 3.0.2, it appears that neither of these actions, nor the closure of the PDE itself, result in a call to the dispose method.

Currently I am ignoring this, but I have a nagging feeling that something is left hanging.

Is this something that is relevant to the current discussion or something else entirely ?

**Comment by jdf** [comment]
Yes. Java-mode processing never calls dispose handlers.

>

-- 
Jonathan Feinberg  jdf@pobox.com  http://MrFeinberg.com/

**Comment by leadedge** [comment]
This is interesting and might give a clue for the Processing developers.

3.0.1, 3.0 and 2.2.1 do call the dispose method. I have checked by putting an infobox in the dispose function. But 3.0.2 does not. At least prior to 3.0.2 it half worked, so the question is what has changed. This could lead to finding a solution.

Also the long standing problem remains that dispose is only called by closing the window with the [x] or by pressing escape. Using the stop button does not call the dispose method.

**Comment by codeanticode** [comment]
@leadedge good lead, thanks. Actually, I did some tests and was wondering if that was actually the case.

The problem with dispose happens with all renderers or only with P2D/P3D? A few months back I committed [this change](https://github.com/processing/processing/commit/727c98a43ab990d16a70db3fcb17eeab30176fca), in order to fix https://github.com/processing/processing/issues/4156. I wonder if this introduced the regression in 3.0.2, at least for the GL renderers.

**Comment by leadedge** [comment]
OK, it gets more interesting.

I set up a test sketch and registered a dispose method in the sketch as well as an exit over-ride, plus a dispose method in the library.

```
//
//            SpoutTest
//
import spout.*;
Spout spout;

void setup() {
  // size(640, 360);
  size(640, 360, P3D);
  spout = new Spout(this);
  registerMethod("dispose", this);
} 

void draw()  { 
    background(0, 90, 100);
}

public void exit() {
  println("Sketch exit method");
  super.exit();
}

public void dispose() {
  println("Sketch dispose method");
  super.dispose();
}
```

I tested with and without the library.

Processing 3.0.2
size(640, 360, P3D) (or P2D)
Close window with [X] - Exit IS called, Dispose is NOT
PDE stop button - nothing called

size(640, 360) or no Size statement at all
Close window with [X] - Exit IS called, Dispose IS called twice
PDE stop button - nothing called

size(640, 360, P3D) + Import a library + create a library object
Close window with [X] - Exit IS called, Dispose is NOT
PDE stop button - nothing called

Processing 3.0.1
size(640, 360, P3D) (or P2D) : 
Close window with [X] - Exit is NOT called, Dispose is NOT
PDE stop button - nothing called

size(640, 360) or no Size statement at all
Close window with [X] - Exit IS called, Dispose IS called twice
PDE stop button - nothing called

size(640, 360, P3D) + Import a library + create a library object
Close window with [X] - Exit is NOT called, Dispose IS called twice  (first once in the sketch, then library dispose and then again in the sketch)
PDE stop button - nothing called


It seems that with 3.0.2, the exit method is called but subsequently dispose is not and so it never reaches the java code in the library. For 3.0.1, exit is not called so it seems then to allow dispose to be called. Not sure why it is called twice, but I observe that the sketch dispose method is called both before and after the library function.

**Comment by leadedge** [comment]
Update - with 3.0.2 if I do this  -

```
public void exit() {
  println("Sketch exit method");
  super.dispose();
  super.exit();
}
```
The sketch exit function is called and the dispose method in the library is called and then the registered dispose method in the sketch is called, but only once.

**Comment by codeanticode** [comment]
These tests are very informative, very much appreciated! I will see how the dispose/exit logic can be properly fixed, will keep you posted.

**Comment by leadedge** [comment]
OK, I hope this helps to sort it through.

Just now I realized that if I can detect the exit method in the sketch (3.0.2), I can create and call a cleanup function in the library. If exit is not called (3.0.1), the dispose method is caught OK, so I can still clean up. I have tested this idea and it works, so I have a solution is the user closes the window with [X].

A problem remains as before if the stop button is pressed. I looked though the processing code and I am a bit lost with the whole thing. But it seems to me that in "runner.java" the "close" function calls "vm.exit". Does that mean java is terminated before the functions in "PSurfaceJOGL.java" have had a chance to happen? As I say I am out of my depth here and could be talking nonsense.

**Comment by jdf** [comment]
Yep, as I said in my comment 3 days ago, above, Processing's sketch runner exits without calling dispose handlers.

**Comment by leadedge** [comment]
Sorry I am slow to come to grips with this. Perhaps the question is whether something can be done before calling "vm.exit(0);". Maybe it is not possible at this point. It is something for the developers to know.

**Comment by jdf** [comment]
Oh, it's definitely a bug, and one that must be addressed by Processing's developers. I believe @codeanticode is looking into it now.

**Comment by hamoid** [comment]
Is there currently a workaround that can be implemented in a library to detect closing of the sketch?

I created the Video Export library, but it's apparently failing on Windows, if people close the sketch window or press ESC. It works fine when people use the stop button on the IDE. In Linux it seems to work always.

My library needs to be called when shutting down, otherwise the movie file that it creates is corrupted (incomplete). For now I can tell users to try the different closing techniques...

**Comment by jdf** [comment]
You might try a ShutdownHook.

**Comment by hamoid** [comment]
@jdf Thanks for the tip. I tried ShutdownHook, and it made me realize that pressing stop on the IDE was actually not calling dispose(),  but for some reason produced videos still worked on my system (maybe on a quad core with SSD it manages to save the data on time, or maybe my video player plays incomplete videos fine).

Unfortunately ShutdownHook does not catch pressing stop on the IDE.

**Comment by jdf** [comment]
Yes, ultimately, the processing developers have to fix the lifecycle bug.

**Comment by codeanticode** [comment]
@leadedge An update on your report from April 4th. First of all, note that the reason for dispose() to be called twice is because you are registering PApplet's dispose, which you don't need to do, as it should be called by Processing when exiting the sketch.

Using the following code with Processing 3.2.1:

```
void setup() {
  size(640, 360, P3D);
  spout = new Spout(this);
} 

void draw()  { 
    background(0, 90, 100);
}

public void exit() {
  println("Sketch exit method");
  super.exit();
}

public void dispose() {
  println("Sketch dispose method");
  super.dispose();
}
```
Exit sketch with <ESC> key: Exit and Dispose ARE BOTH called once
Close window with [X] : Exit IS called once, Dispose is NOT called

With the latest revision from master, dispose IS also called (once) when closing window with [X], I believe thanks to [this fix](https://github.com/processing/processing/commit/faeb919996b1a0ebb2c07458f9568b76d39d9cda). I removed the extra dispose() call in PSurfaceJOGL with https://github.com/processing/processing/commit/0b3cbab5399ba50e2918c00e35dddfaf152d8d3e. Tested on Mac, Windows, and Linux.

Nothing gets called with stopping the sketch from the editor, but that a [separate issue](https://github.com/processing/processing/issues/4445). I think this issue is taken care of with the recent changes.

**Comment by leadedge** [comment]
Brilliant. This has been worrying me for some time. I will look at this shortly.

So I don't need to register the dispose method. Does that mean the documentation should be changed?

https://github.com/processing/processing/wiki/Library-Basics

**Comment by leadedge** [comment]
I have confirmed exactly as you say with Processing 3.2.1 ((REV 0253) - 19 August 2016).

I am assuming that the latest revision from master is not available as a compiled library yet, so I am keen to see that when available, in which case all is fine for close with [X] or ESC. Then it's only the Stop button problem remaining and hopefully the changes you mention will work.

(BTW in the post above I forgot that previously I had registered "dispose" in the sketch itself for testing which is not necessary, so ignore the comment about documentation.)
